### PR TITLE
#140-inventory

### PR DESCRIPTION
**Pull request by Andriets**

### CODE REVIEW

File: EventsExpress.Core/Services/EventService.cs [hunk 1]
```diff
@@ -233,13 +234,13 @@ public async Task<OperationResult> Edit(EventDTO e)
...
             _mapper.Map<EventDTO>(_db.EventRepository
-                .Get("Photo,Owner.Photo,City.Country,Categories.Category,Visitors.User.Photo")
+                .Get("Photo,Owner.Photo,City.Country,Categories.Category,Visitors.User.Photo,Inventories.UnitOfMeasuring")
                 .FirstOrDefault(x => x.Id == eventId));
 
         public IEnumerable<EventDTO> GetAll(EventFilterViewModel model, out int count)
         {
             var events = _db.EventRepository
-                .Get("Photo,Owner.Photo,City.Country,Categories.Category,Visitors");
+                .Get("Photo,Owner.Photo,City.Country,Categories.Category,Visitors,Inventories.UnitOfMeasuring");
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/EventService.cs]: Why do we need to get all inventory for all items? Do we display it on list somewhere?

File: EventsExpress.Core/Services/InventoryService.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            IUnitOfWork unitOfWork,
+            IMapper mapper)
+        {
+            _db = unitOfWork;
+            _mapper = mapper;
+        }
+
+        public async Task<OperationResult> AddInventar(Guid eventId, InventoryDTO inventoryDTO)
+        {
+            var ev = _db.EventRepository.Get("Inventories").FirstOrDefault(e => e.Id == eventId);
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/InventoryService.cs]: If you are checking for existance of event - consider not joining inventory here.

File: EventsExpress.Core/Services/UnitOfMeasuringService.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            catch (Exception ex)
+            {
+                return new OperationResult(false, "Something went wrong " + ex.Message, string.Empty);
+            }
+        }
+
+        public ICollection<UnitOfMeasuringDTO> GetAll()
+        {
+            var entities = _db.UnitOfMeasuringRepository.Get("Inventories").ToList();
+            if (entities == null)
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/UnitOfMeasuringService.cs:67]: Do we really need to pull all inventories here? Let's remove this include.

File: EventsExpress.Core/Services/UnitOfMeasuringService.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                return _mapper.Map<ICollection<UnitOfMeasuring>, ICollection<UnitOfMeasuringDTO>>(entities);
+            }
+        }
+
+        public UnitOfMeasuringDTO GetById(Guid unitOfMeasuringId)
+        {
+            var entity = _db.UnitOfMeasuringRepository.Get(unitOfMeasuringId);
+            if (entity == null)
+            {
+                //throw new NotImplementedException("Object not found!");
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/UnitOfMeasuringService.cs]: Remove commented code.

File: EventsExpress.Db/EF/AppDbContext.cs [hunk 5]
```diff
@@ -37,7 +38,13 @@ public AppDbContext(DbContextOptions<AppDbContext> options)
 
         public DbSet<Country> Countries { get; set; }
 
-        public DbSet<EventStatusHistory> EventStatusHistory { get; set; }
+		public DbSet<EventStatusHistory> EventStatusHistory { get; set; }
```
Comment (iuriikhrystiuk) [inline EventsExpress.Db/EF/AppDbContext.cs]: Consider the formatting.

File: EventsExpress/ClientApp/src/components/event/event-item-view.js [hunk 6]
```diff
@@ -189,8 +205,10 @@ export default class EventItemView extends Component {
             user,
             visitors,
             country,
-            city
+            city,
+            inventories
         } = this.props.event.data;
+        console.log(this.props);
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/components/event/event-item-view.js]: Remove this console.log statement.

File: EventsExpress/ClientApp/src/components/event/event-item-view.js [hunk 7]
```diff
@@ -272,6 +290,53 @@ export default class EventItemView extends Component {
                                 <div className="text-box overflow-auto shadow p-3 mb-5 mt-2 bg-white rounded">
                                     {description}
                                 </div>
+                                <div className="text-box overflow-auto shadow p-3 mb-5 mt-2 bg-white rounded">
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/components/event/event-item-view.js:284]: Please move this inventory display to separate component.

File: EventsExpress/ClientApp/src/components/helpers/helpers.js [hunk 8]
```diff
@@ -49,6 +49,42 @@ export const validate = values => {
         }
     });
 
+    if (!values.inventories || !values.inventories.length) {
+
+    } else {
+        console.log(values);
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/components/helpers/helpers.js]: Remove console.log statement.

File: EventsExpress/ClientApp/src/components/helpers/helpers.js [hunk 9]
```diff
@@ -49,6 +49,42 @@ export const validate = values => {
         }
     });
 
+    if (!values.inventories || !values.inventories.length) {
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/components/helpers/helpers.js]: Don't do empty if statements. Invert the check if required.

File: EventsExpress/ClientApp/src/components/helpers/helpers.js [hunk 10]
```diff
@@ -49,6 +49,42 @@ export const validate = values => {
...
+                inventoriesErrors.needQuantity = 'Can not be negative';
+                inventoriesArrayErrors[index] = inventoriesErrors;
+            }
+            if (!item || !item.unitOfMeasuring) {
+                inventoriesErrors.unitOfMeasuring = {};
+                inventoriesErrors.unitOfMeasuring.id = 'Required';
+                inventoriesArrayErrors[index] = inventoriesErrors;
+            }
+        })
+        console.log(inventoriesArrayErrors);
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/components/helpers/helpers.js]: Remove console.log statement.

File: EventsExpress/ClientApp/src/components/inventory/inventory.js [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    handleOnClickCaret() {
+        this.setState(state => ({
+            isOpen: !state.isOpen
+        }));
+    }
+
+    render() {
+        return (
+            <div>
+                <div className='d-flex justify-content-start align-items-center'>
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/components/inventory/inventory.js:88]: This code is duplicated between display and edit. Consider extracting it to a separate component.

File: EventsExpress/ClientApp/src/store/initialState.js [hunk 12]
```diff
@@ -36,7 +36,11 @@ const initialState = {
         isRegisterSuccess: false,
         registerError: null
     },
-    change_avatar: {
+    unitsOfMeasuring: {
+        isPending: true,
+        isError: false,
+        units: []
+    },    change_avatar: {
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/store/initialState.js]: Starte new piece of state at the new line.

File: EventsExpress/DTO/EventDto.cs [hunk 13]
```diff
@@ -37,5 +37,7 @@ public class EventDto
         public IEnumerable<CategoryDto> Categories { get; set; }
 
         public IEnumerable<UserPreviewDto> Visitors { get; set; }
+
+        public IEnumerable<InventoryDto> Inventories { get; set; }
```
Comment (iuriikhrystiuk) [inline EventsExpress/DTO/EventDto.cs:41]: If this model is used in a preview only - I am not sure that we need inventories here. Please double check and remove it if not required.

**Review by iuriikhrystiuk — CHANGES_REQUESTED** [review]

File: EventsExpress.Core/Services/EventService.cs (hunk 1 above)
Comment (Andriets) [reply EventsExpress.Core/Services/EventService.cs]: Solved

File: EventsExpress.Core/Services/InventoryService.cs (hunk 2 above)
Comment (Andriets) [reply EventsExpress.Core/Services/InventoryService.cs]: Solved

File: EventsExpress.Core/Services/UnitOfMeasuringService.cs (hunk 3 above)
Comment (Andriets) [reply EventsExpress.Core/Services/UnitOfMeasuringService.cs:67]: Solved

File: EventsExpress.Core/Services/UnitOfMeasuringService.cs (hunk 4 above)
Comment (Andriets) [reply EventsExpress.Core/Services/UnitOfMeasuringService.cs]: Solved

File: EventsExpress.Db/EF/AppDbContext.cs (hunk 5 above)
Comment (Andriets) [reply EventsExpress.Db/EF/AppDbContext.cs]: Solved

File: EventsExpress/ClientApp/src/components/event/event-item-view.js (hunk 6 above)
Comment (Andriets) [reply EventsExpress/ClientApp/src/components/event/event-item-view.js]: Solved

File: EventsExpress/ClientApp/src/components/event/event-item-view.js (hunk 7 above)
Comment (Andriets) [reply EventsExpress/ClientApp/src/components/event/event-item-view.js:284]: Solved

File: EventsExpress/ClientApp/src/components/helpers/helpers.js (hunk 8 above)
Comment (Andriets) [reply EventsExpress/ClientApp/src/components/helpers/helpers.js]: Solved

File: EventsExpress/ClientApp/src/components/helpers/helpers.js (hunk 9 above)
Comment (Andriets) [reply EventsExpress/ClientApp/src/components/helpers/helpers.js]: Solved

File: EventsExpress/ClientApp/src/components/helpers/helpers.js (hunk 10 above)
Comment (Andriets) [reply EventsExpress/ClientApp/src/components/helpers/helpers.js]: Solved

File: EventsExpress/ClientApp/src/components/inventory/inventory.js (hunk 11 above)
Comment (Andriets) [reply EventsExpress/ClientApp/src/components/inventory/inventory.js:88]: Solved

File: EventsExpress/ClientApp/src/store/initialState.js (hunk 12 above)
Comment (Andriets) [reply EventsExpress/ClientApp/src/store/initialState.js]: Solved

File: EventsExpress/DTO/EventDto.cs (hunk 13 above)
Comment (Andriets) [reply EventsExpress/DTO/EventDto.cs:41]: EventDto used in edit method only

**Review by iuriikhrystiuk — APPROVED** [review]

**Pull request merged**